test/base-fork/SparkVault.t.sol

```diff
@@ -0,0 +1,221 @@
+        assertApproxEqAbs(asset.balanceOf(address(this)),       state.assetThis,        tolerance, "assetThis");
+        assertApproxEqAbs(asset.balanceOf(address(almProxy)),   state.assetAlm,         tolerance, "assetAlm");
+        assertApproxEqAbs(asset.balanceOf(address(sparkVault)), state.assetVault,       tolerance, "assetVault");
+        assertApproxEqAbs(sparkVault.balanceOf(address(this)),  state.vaultThis,        tolerance, "vaultThis");
+        assertApproxEqAbs(sparkVault.totalAssets(),             state.vaultTotalAssets, tolerance, "vaultTotalAssets");
+        assertApproxEqAbs(sparkVault.totalSupply(),             state.vaultTotalSupply, tolerance, "vaultTotalSupply");
+    }
+
+    function _assertTestState(TestState memory state) internal view {
+        _assertTestState(state, 0);
+    }
+}
+
+contract ForeignControllerTakeFromSparkVaultFailureTests is ForeignControllerTakeFromSparkVaultTestBase {
+
+    function test_takeFromSparkVault_notRelayer() external {
+        vm.expectRevert(abi.encodeWithSignature(
+            "AccessControlUnauthorizedAccount(address,bytes32)",
+            address(this),
+            RELAYER
+        ));
+        // >> Action
+        foreignController.takeFromSparkVault(address(sparkVault), 1e18);
+    }
+
+    function test_takeFromSparkVault_zeroMaxAmount() external {
+        vm.startPrank(Base.SPARK_EXECUTOR);
+        rateLimits.setRateLimitData(key, 0, 0);
+        vm.stopPrank();
+
+        // >> Prank
+        vm.prank(relayer);
+        vm.expectRevert("RateLimits/zero-maxAmount");
+        // >> Action
+        foreignController.takeFromSparkVault(address(sparkVault), 1e18);
+    }
+
+    function test_takeFromSparkVault_rateLimitBoundary() external {
+        deal(address(asset), address(this), 10_000_000e18);
+        asset.approve(address(sparkVault), 10_000_000e18);
+        sparkVault.mint(10_000_000e18, address(this));
+
+        vm.startPrank(Base.SPARK_EXECUTOR);
+        rateLimits.setRateLimitData(key, 10_000_000e18, uint256(10_000_000e18) / 1 days);
+        vm.stopPrank();
+
+        // >> Prank
+        vm.prank(relayer);
+        vm.expectRevert("RateLimits/rate-limit-exceeded");
+        // >> Action
+        foreignController.takeFromSparkVault(address(sparkVault), 10_000_000e18 + 1);
+
+        vm.prank(relayer);
+        foreignController.takeFromSparkVault(address(sparkVault), 10_000_000e18);
+    }
+
+}
+
+contract ForeignControllerTakeFromSparkVaultTests is ForeignControllerTakeFromSparkVaultTestBase {
+
+    function test_takeFromSparkVault_rateLimited() external {
+        deal(address(asset), address(this), 10_000_000e18);
+        asset.approve(address(sparkVault), 10_000_000e18);
+        sparkVault.mint(10_000_000e18, address(this));
+
+        // >> start Prank
+        vm.startPrank(relayer);
+
+        _assertTestState(TestState({
+            rateLimit:        1_000_000e18,
```

Comment:
Can update all of these tests to use the format I used in the value accrual tests: https://github.com/sparkdotfi/spark-vaults-v2/blob/99583d66df7df1698a3f2d2568028e5bb9bcbfde/test/ValueAccrual.t.sol#L118

Only updating the state that changes, will help make things more succinct